Sources/SwiftGRPCNIO/HTTP1ToRawGRPCServerCodec.swift

```diff
@@ -27,111 +27,200 @@ public enum RawGRPCServerResponsePart {
-        }
+    if case .ignore = inboundState { return }
+
+    do {
+      switch self.unwrapInboundIn(data) {
+      case .head(let requestHead):
+        inboundState = try processHead(ctx: ctx, requestHead: requestHead)
+
+      case .body(var body):
+        inboundState = try processBody(ctx: ctx, body: &body)
+
+      case .end(let trailers):
+        inboundState = try processEnd(ctx: ctx, trailers: trailers)
       }
+    } catch {
+      ctx.fireErrorCaught(error)
+      inboundState = .ignore
+    }
+  }
+
+  func processHead(ctx: ChannelHandlerContext, requestHead: HTTPRequestHead) throws -> InboundState {
+    guard case .expectingHeaders = inboundState else {
+      throw StateMachineError("expecteded state .expectingHeaders, got \(inboundState)")
+    }
+
+    ctx.fireChannelRead(self.wrapInboundOut(.head(requestHead)))
 
-    case .end(let trailers):
-      if let trailers = trailers {
-        //! FIXME: Better handle this error.
-        print("unexpected trailers received: \(trailers)")
-        return
+    return .expectingBody(.expectingCompressedFlag)
+  }
+
+  func processBody(ctx: ChannelHandlerContext, body: inout ByteBuffer) throws -> InboundState {
+    guard case .expectingBody(let bodyState) = inboundState else {
+      throw StateMachineError("expecteded state .expectingBody(_), got \(inboundState)")
+    }
+
+    return .expectingBody(try processBodyState(ctx: ctx, initialState: bodyState, messageBuffer: &body))
+  }
+
+  func processBodyState(ctx: ChannelHandlerContext, initialState: InboundState.Body, messageBuffer: inout ByteBuffer) throws -> InboundState.Body {
+    var bodyState = initialState
+
+    // Iterate over all available incoming data, trying to read length-delimited messages.
+    // Each message has the following format:
+    // - 1 byte "compressed" flag (currently always zero, as we do not support for compression)
+    // - 4 byte signed-integer payload length (N)
+    // - N bytes payload (normally a valid wire-format protocol buffer)
+    while true {
+      switch bodyState {
+      case .expectingCompressedFlag:
+        guard let compressionFlag: Int8 = messageBuffer.readInteger() else { return .expectingCompressedFlag }
+
+        // TODO: Add support for compression.
+        guard compressionFlag == 0 else { throw GRPCStatus.unsupportedCompression }
+
+        bodyState = .expectingMessageLength
+
+      case .expectingMessageLength:
+        guard let messageLength: UInt32 = messageBuffer.readInteger() else { return .expectingMessageLength }
+        bodyState = .receivedMessageLength(messageLength)
+
+      case .receivedMessageLength(let messageLength):
+        // We need to account for messages being spread across multiple `ByteBuffer`s so buffer them
+        // into `buffer`. Note: when messages are contained within a single `ByteBuffer` we're just
+        // taking a slice so don't incur any extra writes.
+        guard messageBuffer.readableBytes >= messageLength else {
```

Comment:
Shouldn't this take into account the amount of data already read into `buffer`, if available?

Example:

Frame A contains message length (100) plus 50 bytes of the message. After receiving frame A, we will be in state `.receivedMessageLength(100)`, with 50 bytes in `buffer`.
If we now receive frame B with the remaining 50 bytes, this `guard` will bail, even though we can assemble the remaining data by concatenating `buffer` and `messageBuffer`. Let me know if I'm missing something.

Reply:
We update the state with the number of bytes we still need (see L153). 

I can see how the naming leads to confusion though, perhaps renaming `receivedMessageLength` to `expectingBytes` and `messageLength` to `requiredBytes` would help?

```swift
case .expectingBytes(let requiredBytes):
  // ...
```

Reply:
Ah, thanks for the clarification! I indeed missed that. We should definitely do the renames you suggested, but maybe use something to the tune of `expectingMoreMessageBytes` and `bytesOutstanding`?

Reply:
No problem -- that naming works for me!